java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -532,14 +570,18 @@ class DatabaseImpl(
             put("collection_id", collectionId)
         }
         // TODO(#4889): Don't do this one-by-one.
-        val valueIds = if (isPrimitiveType(typeId)) {
-            elements.map { getPrimitiveValueId(it as Referencable, typeId, db) }
-        } else {
-            elements.map {
-                require(it is Reference) {
-                    "Expected element in collection to be a Reference but was $it."
+        val valueIds = when {
+            isPrimitiveType(typeId) && fieldClass == FieldClass.Collection ->
+                elements.map { getPrimitiveValueId(it as Referencable, typeId, db) }
+            isPrimitiveType(typeId) && fieldClass == FieldClass.List ->
+                elements.map { getPrimitiveValueId(it as Referencable, typeId, db) }
```

Comment:
looks like you are doing the same thing for both lists and collections, so why check the FieldClass?

Reply:
If the fieldType is Singleton then I need to do the else path. I'll rearrange

Reply:
no, that's not right. I think you're right, I can just check isPrimitiveType